src/uct/tcp/tcp_ep.c

```diff
@@ -984,8 +1206,12 @@ static unsigned uct_tcp_ep_progress_am_rx(uct_tcp_ep_t *ep)
                    ((ep->rx.length - ep->rx.offset) >= sizeof(*hdr)));
 
         /* do partial receive of the remaining user data */
-        hdr         = UCS_PTR_BYTE_OFFSET(ep->rx.buf, ep->rx.offset);
-        recv_length = hdr->length - (ep->rx.length - ep->rx.offset - sizeof(*hdr));
+        hdr             = UCS_PTR_BYTE_OFFSET(ep->rx.buf, ep->rx.offset);
+        if (hdr->length > (ep->rx.length - ep->rx.offset - sizeof(*hdr))) {
+            recv_length = hdr->length - (ep->rx.length - ep->rx.offset - sizeof(*hdr));
```

Comment:
use local var for `hdr->length - (ep->rx.length - ep->rx.offset - sizeof(*hdr))`